examples/ppi_unsup.py

```diff
@@ -0,0 +1,119 @@
+import os.path as osp
+
+import torch
+import torch.nn.functional as F
+from sklearn.linear_model import SGDClassifier
+from sklearn.metrics import f1_score
+from sklearn.multioutput import MultiOutputClassifier
+
+from torch_geometric.data import Batch
+from torch_geometric.datasets import PPI
+from torch_geometric.loader import DataLoader, LinkNeighborLoader
+from torch_geometric.nn import SAGEConv
+
+EPS = 1e-15
+
+path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'PPI')
+train_dataset = PPI(path, split='train')
+val_dataset = PPI(path, split='val')
+test_dataset = PPI(path, split='test')
+
+# Training loaders (one datapoint corresponds to a single node)
+train_data = Batch.from_data_list(train_dataset)
+loader = LinkNeighborLoader(train_data, batch_size=2048, neg_sampling_ratio=.5,
+                            num_neighbors=[10, 10])
+
+# Evaluation loaders (one datapoint corresponds to a graph)
+train_loader = DataLoader(train_dataset, batch_size=2, shuffle=False)
+val_loader = DataLoader(val_dataset, batch_size=2, shuffle=False)
+test_loader = DataLoader(test_dataset, batch_size=2, shuffle=False)
+
+
+class GraphSAGE(torch.nn.Module):
+    def __init__(self, in_channels, hidden_channels, num_layers):
+        super().__init__()
+        self.num_layers = num_layers
+        self.convs = torch.nn.ModuleList()
+        for i in range(num_layers):
+            in_channels = in_channels if i == 0 else hidden_channels
+            self.convs.append(SAGEConv(in_channels, hidden_channels))
+
+    def forward(self, x, edge_index):
+        for i, conv in enumerate(self.convs):
+            x = conv(x, edge_index)
+            if i != self.num_layers - 1:
+                x = x.relu()
+                x = F.dropout(x, p=0.5, training=self.training)
+        return x
+
+
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+model = GraphSAGE(train_dataset.num_features, 64, 2).to(device)
+optimizer = torch.optim.Adam(model.parameters(), lr=0.005)
+
+
+def train():
+    model.train()
+
+    total_loss = 0
+    for data in loader:
+        data = data.to(device)
+        optimizer.zero_grad()
+        out = model(data.x, data.edge_index)[:data.edge_label.shape[0]]
```

Comment:
There is no need for this separation anymore. You should be able to simply do:
```python
out = model(data.x, data.edge_index)
link_pred = (out[data.edge_label_index[0]] * out[data.edge_label_index[1]]).sum(dim=-1)
loss = F.bce_loss(out, data.edge_label)
```

Reply:
Thanks @rusty1s, did you mean `F.bce_loss(out, data.edge_label)`

Reply:
Yes, sorry :)